### DISCUSSION TITLE
Noob question: Type check for HyExpression?

### DISCUSSION THREAD

**Discussion by nicoszerman** [hylang/hy#2152] [Q&A] [answered]
Hi! I am learning Lisp and I definitely think that Hy is the way to go for me.

I am writing a macro and would like to recursively iterate over the code passed. I want to check if a variable contains code I should iterate on, or not. At the moment I am using `(if (iterable? code)` but `iterable?` returns true for more than just code. I was wondering, is there any type check for Hy code? Something like `(HyExpression? code)` ?

Thank you so much in advance.

**Comment by nicoszerman** [comment]
My solution `(isinstance code HyExpression)`

**Comment by Kodiologist** [comment, marked as answer]
`hy.models.Expression` (formerly `HyExpression`) is only one of several types that Hy uses for its code. The supertype that everything else inherits from is `hy.models.Object`. Note however that checking `(isinstance code hy.models.Object)` won't catch objects that could be promoted to models, such as `5`, which would be promoted to `hy.models.Integer` in e.g. `(hy.eval 5)`. The most general approach is to call `hy.as-model` and check if it raises an error.

**Reply by nicoszerman** [reply]
I'm learning Hy. With code I meant parentheses with symbols inside `(code code ... code)`, so not numbers, not lists []. Is the best approach to use `(isinstance code HyExpression)`? By any chance do you know how this is done in other dialects of Lisp?

**Reply by Kodiologist** [reply]
Yeah, parentheses create `Expression`s. In most Lisps, they create lists made out of cons cells, so you'd use e.g. [CONSP](http://clhs.lisp.se/Body/f_consp.htm) in Common Lisp.

**Reply by nicoszerman** [reply]
Awesome that makes sense thank you so much. Shouldn't there be a Hy equivalent to CONSP ? It seems such a useful method. And having a specific method `(expression? foo)` equivalent to `(isinstance foo HyExpression)` can help compilers in the future optimize Hy code.

**Reply by Kodiologist** [reply]
We used to have a whole swarm of such predicates, but I killed them off (#2095) because they're rarely used, they save just a few characters, and it wasn't clear where to draw the line as to what to include. And since Hy has Python's dynamic semantics by design, it's unlikely that optimizations of function calls like that are ever going to happen.
